frappe/utils/safe_exec.py

```diff
@@ -188,6 +193,40 @@ def get_safe_globals():
 
 	return out
 
+def safe_enqueue(function, queue='default', job_name=None, **kwargs):
+	'''
+		Enqueue function to be executed using a background worker if not already queued
+
+		:param function: whitelised function or server script of type API
+		:param queue: should be either long, default or short
+		:param job_name: used to identify an enqueue call, used to prevent duplicate calls
+	'''
+
+	job_name = job_name or function
+
+	site = frappe.local.site
+	queued_jobs = get_jobs(site=site, queue=queue, key='job_name').get(site) or []
+
+	if job_name not in queued_jobs:
+		return enqueue(
+			'frappe.utils.safe_exec.call_whitelisted_function',
+			job_name=job_name, queue=queue, function=function, **kwargs
+		)
+
+def call_whitelisted_function(function, **kwargs):
+	'''
+	Executes the function if it is whitelisted or a server script of type API
+	'''
+
+	# update form_dict, to use as `kwargs` in frappe.call
+	form_dict = getattr(frappe.local, 'form_dict', {})
+	frappe.local.form_dict = frappe._dict(kwargs or {})
```

Comment:
Should `form_dict` be overwritten or updated with `kwargs`?

Reply:
What's the use-case for overwriting it?

Reply:
No, I was thinking the current `form_dict` should just be updated with `kwargs`.

Can values set in `form_dict` in one script be used in another script?
If not, then overwriting and restoring is fine.

Reply:
Sorry, I understand what you mean now. Yes, these values _can_ be updated in the script itself. I guess we can update the form_dict, as long as we are restoring it back.

Reply:
Then there is another issue: what if the inner function is updating `form_dict` in a way that outer function benefits?

Reply:
https://github.com/frappe/frappe/pull/15213/commits/83802dac7fcd44c22e5dac2136d90cad3c2bcc6f

Reply:
Part of me thinks we should just introduce breaking changes to design it properly and consistently.